hack/ci/custom-linters/setuplognilerrorcheck/setuplognilerrorcheck.go

```diff
@@ -0,0 +1,83 @@
+package setuplognilerrorcheck
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/format"
+	"go/token"
+	"go/types"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+var SetupLogNilErrorCheck = &analysis.Analyzer{
+	Name: "setuplognilerrorcheck",
+	Doc:  "Detects improper usage of logger.Error() calls, ensuring the first argument is a non-nil error.",
+	Run:  runSetupLogNilErrorCheck,
+}
+
+func runSetupLogNilErrorCheck(pass *analysis.Pass) (interface{}, error) {
+	for _, f := range pass.Files {
+		ast.Inspect(f, func(n ast.Node) bool {
+			callExpr, ok := n.(*ast.CallExpr)
+			if !ok {
+				return true
+			}
+
+			// Ensure function being called is logger.Error
+			selectorExpr, ok := callExpr.Fun.(*ast.SelectorExpr)
+			if !ok || selectorExpr.Sel.Name != "Error" {
+				return true
+			}
+
+			// Ensure receiver (logger) is identified
+			ident, ok := selectorExpr.X.(*ast.Ident)
+			if !ok {
+				return true
+			}
+
+			// Verify if the receiver is logr.Logger
+			obj := pass.TypesInfo.ObjectOf(ident)
+			if obj == nil {
+				return true
+			}
+
+			if named, ok := obj.Type().(*types.Named); ok {
+				if named.Obj().Pkg() != nil && named.Obj().Pkg().Path() == "github.com/go-logr/logr" && named.Obj().Name() == "Logger" {
+					if len(callExpr.Args) > 0 {
+						firstArg := callExpr.Args[0]
+						suggestedError := "errors.New(\"kind error (i.e. configuration error)\")"
+						suggestedMessage := "\"error message describing the failed operation\""
+
+						// Extract error message for the suggestion
+						if len(callExpr.Args) > 1 {
+							if msgArg, ok := callExpr.Args[1].(*ast.BasicLit); ok && msgArg.Kind == token.STRING {
+								suggestedMessage = msgArg.Value
+							}
+						}
+
+						// Get the actual source code line where the issue occurs
+						var srcBuffer bytes.Buffer
+						if err := format.Node(&srcBuffer, pass.Fset, callExpr); err == nil {
+							sourceLine := srcBuffer.String()
+
+							// Check if the first argument of the error log is nil
+							if arg, ok := firstArg.(*ast.Ident); ok && arg.Name == "nil" {
+								pass.Reportf(callExpr.Pos(),
+									"Incorrect usage of 'logger.Error(nil, ...)'. The first argument must be a non-nil 'error'. "+
+										"Passing 'nil' results in silent failures, making debugging harder.\n\n"+
+										"\U0001F41B **What is wrong?**\n"+
+										fmt.Sprintf("   %s\n\n", sourceLine)+
+										"\U0001F4A1 **How to solve? Return the error, i.e.:**\n"+
+										fmt.Sprintf("   logger.Error(%s, %s, \"key\", value)\n\n", suggestedError, suggestedMessage))
+							}
+						}
+					}
+				}
+			}
```

Comment:
nit: is there a way to decrease the nesting by inverting the checks and returning early?

Reply:
Done 👍 Good point